x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts

```diff
@@ -84,6 +103,52 @@ export const importRules = async ({
               return null;
             }
 
+            if (!allowPrebuiltRules && parsedRule.immutable) {
+              resolve(
+                createBulkErrorObject({
+                  statusCode: 400,
+                  message: i18n.translate(
+                    'xpack.securitySolution.detectionEngine.rules.importPrebuiltRulesUnsupported',
+                    {
+                      defaultMessage:
+                        'Importing prebuilt rules is not supported. To import this rule as a custom rule, remove its "immutable" property try again.',
+                    }
+                  ),
+                  ruleId: parsedRule.rule_id,
+                })
+              );
+
+              return null;
+            }
+
+            if (allowPrebuiltRules) {
+              // TODO how do we differentiate this case from a non-prebuilt import?
+              if (!ruleImportIsPrebuilt(parsedRule)) {
+                resolve(
+                  createBulkErrorObject({
+                    statusCode: 400,
+                    message: i18n.translate(
+                      'xpack.securitySolution.detectionEngine.rules.cannotImportPrebuiltRuleWithoutVersion',
+                      {
+                        defaultMessage: 'Prebuilt rules must specify a "version" to be imported.',
+                      }
+                    ),
+                    ruleId: parsedRule.rule_id,
+                  })
+                );
+
+                return null;
+              }
+
+              const ruleSource = calculateRuleSourceForImport({
+                rule: parsedRule,
+                prebuiltRuleAssets,
+                installedRuleIds,
+              });
+
+              parsedRule.rule_source = ruleSource;
+            }
+
```

Comment:
Passing `parsedRule` to `await detectionRulesClient.importRule()` won't be enough to import a customized prebuilt rule with the right `rule_source` and `immutable` fields. Even though the calculation of ruleSource here is right, when we pass `parsedRule` to `await detectionRulesClient.importRule()` the `rule_source` and `immutable` will be recreated and the values from `parsedRule` ignored; this is the current functionality of the `importRule` method, since when we created the `detectionRulesClient` importing rules only worked for custom rules.

So we need to modify `detectionRulesClient.importRule` (x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/methods/import_rule.ts) so that the calculated value for `rule_source` and `immutable` are maintained when we create the new rule being imported, or when we updated the rule being overwritten on import.

But if you take a look at that file now, you'll see that:
- if overwriting an existing rule, before calling `await rulesClient.update` we call `applyRuleUpdate`, which internally calls `calculateRuleSource` passing the existing rule with its updates as the `nextRule`, which is wrong since it will contain a wrong values for `immutable` (`calculateRuleSource` still relies on the existing rule's value for the `immutable` field, while we want to completely ignore the field and recalculate it and `rule_source` from scratch).
- if creating a new rule, we call `createRule`, which internally calls `applyRuleDefaults`, which calculates `rule_source` only from the `immutable` property with the `convertImmutableToRuleSource` method. So also here we need to change this logic and rely on the data that you calculate beforehand in `calculateRuleSourceForImport` instead.

Reply:
This is the last remaining logical piece, I believe. I took a first pass at this but the relevant functions (`applyRuleUpdate`, `createRule`) aren't typed to expect `rule_source` or `immutable` values: they expect `RuleUpdateProps` or `RuleCreateProps`.

I'm going to revisit the RFC's discussion about refactoring all these create/update methods and think about this overnight. I'm currently leaning toward either pushing those calculations to this level, or just writing an entirely new code path if the FF is enabled (since a lot of this logic seems unnecessary once we know that these fields will be calculated/available). I'll update this thread tomorrow with my findings, and we can discuss on Monday.